### ISSUE TITLE
Issue with links in facebook email after cleaning

### ISSUE THREAD

**Issue by foxinushka** [htacg/tidy-html5#551]
Tidy brakes next html:
https://gist.github.com/foxinushka/7fc2dbce198bcfa02379b03da77dbd29

It is a part of Facebook email.
Tidy extracts 
```html
<a href=...> ... </a>
```
 content to separate block.

Cleaned html:
https://gist.github.com/foxinushka/8f880497c59cbfbff4e0626f30c2827e

I've tested next versions:
5.3.16 (current version of Tidy in our product)
5.4.0
5.5.20
The result is the same.

**Comment by foxinushka** [comment]
I've just tested 4.9.11 (the old version, which we were using before 5.3.16).
I worked better. It had leaved content inside 
```html
<a href=...> ... </a>
```
block. But I had duplicated link.

Cleaned html in 4.9.11:
https://gist.github.com/foxinushka/79da3052490539a2fcbda4727a02ee23

**Comment by foxinushka** [comment]
This behavior was changed in 86f626cd674d09c9215c59c4cdad1d4e7887e00b.

**Comment by foxinushka** [comment]
So, I've debuged, that problem now with tags adjustments, because this content is HTML 4.01. Where can I read about tag a, if it can't be a block in HTML 4.01?

```C
            if (lexer->doctype != VERS_HTML5)
                    {
                        /*\
                         *  Back to legacy HTML4 mode for -
                         *  Issue #167 & #169 - TidyTag_A
                         *  Issue #196        - TidyTag_CAPTION
                         *  others?
                        \*/ 
                        TY_(AdjustTags)(doc); /* Dynamically modify the tags table  */
                    }
```

**Comment by geoffmcl** [comment]
@foxinushka thanks for the issue, and the exploration of the tidy versions, commits and code... that is **good** to see... a person trying to help themselves...

> Where can I read about tag a, if it can't be a block in HTML 4.01?

Yes, the `<a>` tag changed - see [a-changes](http://w3c.github.io/html-reference/a.html#a-changes) - where it essentially changed from an inline `phasing-content` to `transparent` ie a block element...

So since your document contains a HTML 4.01 doctype, Tidy will do `TY_(AdjustTags)(doc);`, and treat it as `inline`... and close the `<a>` on reaching the `<center>` tag... with catastrophic results... we no longer have a `facebook` link...

If you change the doctype to `<!DOCTYPE html>`, the transparent block nature of the anchor will produce what you want... I think... try it...

Of course, Tidy will still `bark` about the use of `<center>` and `<font>`, elements removed from html5. And so will the W3C validator!

Am marking this as `Technical Support` until we can narrow down exactly what you expect tidy to do in this case... 

Look forward to further feedback and comments... thanks...

PS: Just so we are sure we are testing the same document have pushed [in_551.html](https://github.com/geoffmcl/tidy-test/blob/master/test/input5/in_551.html), your original, and [in_551-2.html](https://github.com/geoffmcl/tidy-test/blob/master/test/input5/in_551-2.html), with a html5 doctype...

**Comment by foxinushka** [comment]
@geoffmcl thanks for explanation.

> If you change the doctype to <!DOCTYPE html>, the transparent block nature of the anchor will produce what you want... I think... try it...

I can't change doctype to HTML5, because it change content representation in some cases.

I use TidyHtml lib in Spark email client to prepare content for XML parser. (I parse e-mails content to better presentation in Spark).
So I use it to fix HTML and convert it to XHTML.
But in output of my parser I can get HTML content, if it decides that the content should be represented as HTML.

I think about some option, which disables some fixes for a-tags in Tidy. I understand, that the HTML is wrong, but it is part of real e-mail from Facebook, and most browsers work well with it. 

Also I have other issue with this e-mail which follow me to disable this fix:

```
parser.c:1343
 /*\
         *  Issue #307 - an <A> tag to ends any open <A> element
         *  Like #427827 - fixed by Randy Waki and Bjoern Hoehrmann 23 Aug 00
         *  in ParseInline(), fix copied HERE to ParseBlock()
         *  href: http://www.w3.org/TR/html-markup/a.html
         *  The interactive element a must not appear as a descendant of the a element.
        \*/
        if ( nodeIsA(node) && !node->implicit && 
             (nodeIsA(element) || DescendantOf(element, TidyTag_A)) )
        {
```


What can you suggest me to solve this problem?
I can pull request some new options for Tidy process, or make branch in my fork with this 'hacky' options.

**Comment by geoffmcl** [comment]
@foxinushka thanks for the further explanation...

> I can't change doctype to HTML5, because it change content representation in some cases.

That makes things quite difficult. Seems you **want** legacy html4 parsing, oh, **except** for the `<a>` tag! Hmmmm...

Since it seems you can compile **`libTidy`** from source, you could certainly **fork** the source, add you own branch, and add you own option like say `--treat-a-as-html5 yes`, or something, then avoid changing the parser or model of `TidyTag_A` in `AdjustTags`... it would continue to be parsed as html5...

I am sorry `most browsers work well with it` does **not** cut it for Tidy, which is associated with the W3C. Tidy will always have to strive for **doing the right thing**, as far as html specs go, so your above patch, even as an option, may not be accepted to become part of official tidy... but would work for your use case...

In other words, if you learn a bit of `git` magic, you could easily maintain your own **custom** version of the library... that is good, correct use of open source... encouraged even...

Not sure what you are trying to say about `parser.c:1343` - `Issue #307 - an <A> tag to ends any open <A> element`!

There seems **no case** for disabling this. You can **not** have a `link` within a `link`, and **must** mess up in most browsers, since how would they know which `link` to follow `onclick`! Of course, you may get lucky, and get what you want or expect, but it remains **bad** html coding...

But maybe I misunderstand something here?

Anyway, hope this helps... be happy to try to answer any other questions... thanks...

**Comment by foxinushka** [comment]
Thanks, it helped me : )

So, I will make fork of TidyHtml and will use it with my custom options for this cases.
I understand that "most browsers work well with it" it isn't a point for the library, because its purpose to fix HTMLs. I told about this only to explain my situation.

Thanks for great library : )
It helps me a lot in Spark.
